llama_stack/distribution/configure.py

```diff
@@ -174,4 +176,9 @@ def parse_and_maybe_upgrade_config(config_dict: Dict[str, Any]) -> StackRunConfi
 
     config_dict["version"] = LLAMA_STACK_RUN_CONFIG_VERSION
 
+    if not config_dict.get("external_providers_dir", None):
+        config_dict["external_providers_dir"] = EXTERNAL_PROVIDERS_DIR
+        if not os.path.exists(EXTERNAL_PROVIDERS_DIR):
+            os.makedirs(EXTERNAL_PROVIDERS_DIR, exist_ok=True)
```

Comment:
why is a function which is supposed to upgrade a config making a completely unwarranted side-effect by creating a directory?!

Reply:
@ashwinb `~/.llama` in general seems to be created as a side effect: https://github.com/meta-llama/llama-stack/blob/e6bbf8d20b87f46794bda659d8d15b0aa779f218/llama_stack/cli/stack/_build.py#L362
so this is following that precedent just in a new spot (when we render the run config).

I think it would make sense to standardize where/when all of `~/.llama` and its sub-dirs are created (perhaps in a setup.py or something), but this is kind of a consequence of the fact that ~/.llama/providers.d should exist, but I am unsure of what standard place to render this to disk.

Reply:
@cdoern my concern is that _this function_ should not be in the business of making this directory (of course `llama stack build` makes all kinds of changes)

Reply:
@ashwinb do you have a suggestion of where a better place might be to do this creation?

Reply:
We could - `ramalama-stack` actually will do the creation with the method we've proposed: https://github.com/nathan-weinberg/ramalama-stack/blob/write-files/setup.py#L16

We would just to need make sure that all external provider made sure to do this

Reply:
well just do it in the caller of this function, no? my comment is more about the sanity of this function. functions which shouldn't have side-effects should remain so or declare their intent to mutate in their names.

in this case, just do this in the `_build.py` or wherever this config upgrade gets called and we should be fine.

Reply:
@ashwinb just pushed a change, let me know if that makes sense!